model-optimizer/extensions/front/tf/SSliceComplexRolledFFTPackBlockReplacement.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+
+import logging as log
+
+from extensions.middle.InsertLayoutPropagationTransposes import mark_input_as_in_correct_layout
+from extensions.ops.dft import DFT, IDFT
+from extensions.ops.roll import Roll
+from mo.front.common.partial_infer.utils import int64_array
+from mo.front.common.replacement import FrontReplacementSubgraph
+from mo.front.subgraph_matcher import SubgraphMatch
+from mo.front.tf.graph_utils import create_op_with_const_inputs
+from mo.graph.graph import Graph, Node, rename_nodes
+
+
+class SSliceComplexRolledFFTPackBlockReplacement(FrontReplacementSubgraph):
+    enabled = True
+
+    def pattern(self):
+        return dict(
+            nodes=[
+                ('strided_slice_real', dict(op='StridedSlice')),
+                ('strided_slice_imag', dict(op='StridedSlice')),
+                ('complex', dict(op='Complex')),
+                ('roll', dict(op='Roll')),
+                ('fft', dict(op='TFFFT')),
+                ('unroll', dict(op='Roll')),
+                ('real', dict(op='Real')),
+                ('imag', dict(op='Imag')),
+                ('pack', dict(op='Pack')),
+            ],
+            edges=[
+                ('strided_slice_real', 'complex', {'in': 0}),
+                ('strided_slice_imag', 'complex', {'in': 1}),
+                ('complex', 'roll', {'in': 0}),
+                ('roll', 'fft', {'in': 0}),
+                ('fft', 'unroll', {'in': 0}),
+                ('unroll', 'real', {'in': 0}),
+                ('unroll', 'imag', {'in': 0}),
+                ('real', 'pack', {'in': 0}),
+                ('imag', 'pack', {'in': 1}),
+            ])
+
+    def replace_sub_graph(self, graph: Graph, match: [dict, SubgraphMatch]):
+        strided_slice_real = match['strided_slice_real']
+        strided_slice_imag = match['strided_slice_imag']
+
+        real_input = strided_slice_real.in_port(0).get_source().node
+        imag_input = strided_slice_imag.in_port(0).get_source().node
+        if real_input.soft_get('name', real_input.id) != imag_input.soft_get('name', imag_input.id):
+            log.debug('The pattern does not correspond to (i)fftxd with shift. Different inputs.')
+            return
+
+        roll = match['roll']
+        unroll = match['unroll']
+        roll_name = roll.soft_get('name', roll.id)
+        unroll_name = unroll.soft_get('name', unroll.id)
+
+        roll_before = Roll(graph, {}).create_node()
+        roll_after = Roll(graph, {}).create_node()
```

Comment:
Can not we use `Roll` operations from the original graph? I guess, it might reduce code amount in this transformation. Please, tell me if I am wrong

Reply:
If we use `Roll` from the original graph,  we get shape infer failure.